pkg/test/test.go

```diff
@@ -0,0 +1,171 @@
+package test
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"testing"
+
+	petname "github.com/dustinkirkland/golang-petname"
+	testutils "github.com/kudobuilder/kudo/pkg/test/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var testStageRegex = regexp.MustCompile(`^(\d+)-([^.]+)(.yaml)?$`)
+
+// Test contains all of the test stages and the Kubernetes client and other global configuration
+// for a test.
+type Test struct {
+	Stages []*Stage
+	Name   string
+	Dir    string
+
+	Client          client.Client
+	DiscoveryClient discovery.DiscoveryInterface
+	Logger          testutils.Logger
+}
+
+// DeleteNamespace deletes a namespace in Kubernetes after we are done using it.
+func (t *Test) DeleteNamespace(namespace string) error {
+	t.Logger.Log("Deleting namespace:", namespace)
+	return t.Client.Delete(context.TODO(), &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Namespace",
+		},
+	})
+}
+
+// CreateNamespace creates a namespace in Kubernetes to use for a test.
+func (t *Test) CreateNamespace(namespace string) error {
+	t.Logger.Log("Creating namespace:", namespace)
+	return t.Client.Create(context.TODO(), &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Namespace",
+		},
+	})
+}
+
+// TestFactory creates a new Go test that runs a set of test stages.
+func (t *Test) TestFactory() func(*testing.T) {
+	return func(test *testing.T) {
+		t.Logger = testutils.NewTestLogger(test, t.Name)
+
+		test.Parallel()
```

Comment:
Under `Proposal` it says ```within each test, test stages run sequentially``` so I probably would just need more clarification if we are running here the test stage parallel. Right now that's unclear to me and from the comment on line 69 it seems like this could also lead to race conditions.

Reply:
This starts the tests, not the stages. Tests run in parallel, stages are what run sequentially. On line 69 we create a namespace to deploy our resources into for the test so that the tests can run in parallel (though, have not decided how to handle cluster-level resources other than ensuring their uniqueness across a test suite).

Reply:
lgtm!